nixos/lib/make-zfs-image.nix

```diff
@@ -0,0 +1,281 @@
+  tools = lib.makeBinPath (
+    with pkgs; [
+      config.system.build.nixos-enter
+      config.system.build.nixos-install
+      dosfstools
+      e2fsprogs
+      nix
+      parted
+      utillinux
+      zfs
+    ]
+  );
+
+  stringifyProperties = prefix: properties: lib.concatStringsSep " \\\n" (
+    lib.mapAttrsToList
+      (
+        property: value: "${prefix} ${lib.escapeShellArg property}=${lib.escapeShellArg value}"
+      )
+      properties
+  );
+
+  createDatasets =
+    let
+      datasetlist = lib.mapAttrsToList lib.nameValuePair datasets;
+      sorted = lib.sort (left: right: (lib.stringLength left.name) < (lib.stringLength right.name)) datasetlist;
+      cmd = { name, value }:
+        let
+          properties = stringifyProperties "-o" (value.properties or {});
+        in
+          "zfs create -p ${properties} ${poolName}/${name}";
+    in
+      lib.concatMapStringsSep "\n" cmd sorted;
+
+  mountDatasets =
+    let
+      datasetlist = lib.mapAttrsToList lib.nameValuePair datasets;
+      mounts = lib.filter ({ value, ... }: value ? "mount") datasetlist;
+      sorted = lib.sort (left: right: (lib.stringLength left.value.mount) < (lib.stringLength right.value.mount)) datasetlist;
+      cmd = { name, value }:
+        ''
+          mkdir -p /mnt/${lib.escapeShellArg value.mount}
+          mount -t zfs ${poolName}/${name} /mnt/${lib.escapeShellArg value.mount}
+        '';
+    in
+      lib.concatMapStringsSep "\n" cmd sorted;
+
+  unmountDatasets =
+    let
+      datasetlist = lib.mapAttrsToList lib.nameValuePair datasets;
+      mounts = lib.filter ({ value, ... }: value ? "mount") datasetlist;
+      sorted = lib.sort (left: right: (lib.stringLength left.value.mount) > (lib.stringLength right.value.mount)) datasetlist;
+      cmd = { name, value }:
+        ''
+          umount /mnt/${lib.escapeShellArg value.mount}
+        '';
+    in
+      lib.concatMapStringsSep "\n" cmd sorted;
+
+
+  fileSystemsCfgFile =
+    let
+      mountable = lib.filterAttrs (_: value: value ? "mount") datasets;
+    in
+      pkgs.writeText "filesystem-config.nix"
+        (
+          "builtins.fromJSON ''" + (
+            builtins.toJSON {
+              fileSystems = lib.mapAttrs'
+                (
+                  dataset: attrs:
+                    {
+                      name = attrs.mount;
+                      value = {
+                        fsType = "zfs";
+                        device = "${poolName}/${dataset}";
+                      };
+                    }
+                )
+                mountable;
+            }
+          ) + "''"
+        );
+
+  mergedConfig =
+    if configFile == null
+    then fileSystemsCfgFile
+    else
+      pkgs.runCommand "configuration.nix" {}
+        ''
+          (
+            echo '{ imports = ['
+            printf "(%s)\n" "$(cat ${fileSystemsCfgFile})";
+            printf "(%s)\n" "$(cat ${configFile})";
+            echo ']; }'
+          ) > $out
+        '';
```

Comment:
This is a bit weird. Since the datasets are dynamic, the config on disk needs to be dynamic too. It therefore merges the config passed in from https://github.com/NixOS/nixpkgs/pull/106574/files#diff-15d957b64e4bcfadbe6317cb441ec661eacac936af6e918f289cfffbb46444edR77-R89 with the `fileSystems` config generated based on the `datasets` provided.

Reply:
Resolved by synthesizing fileSystems from the datasets argument provided at provision time..